_posts/2021-10-21-segmentation-torch-android/segmentation-torch-android.Rmd

```diff
@@ -0,0 +1,279 @@
+
+![A cat from the Oxford Pet Dataset (@parkhi12a).](images/Abyssinian_47.jpg)
+
+## Train in R
+
+We start by preparing the data.
+
+### Pre-processing and data augmentation
+
+As provided by `torchdatasets`, the [Oxford Pet Dataset](https://www.robots.ox.ac.uk/~vgg/data/pets/) comes with three variants of target data to choose from: the overall class (cat or dog), the individual breed (there are thirty-seven of them), and a pixel-level segmentation with three categories: foreground, boundary, and background. The latter is the default; and it's exactly the type of target we need.
+
+A call to `oxford_pet_dataset(root = dir)` will trigger the initial download:
+
+```{r}
+# need torch > 0.6.1
+# may have to run remotes::install_github("mlverse/torch") depending on when you read this
+library(torch) 
+library(torchvision)
+library(torchdatasets)
+library(luz)
+
+dir <- "~/.torch-datasets/oxford_pet_dataset"
+
+ds <- oxford_pet_dataset(root = dir)
+```
+
+Images (and corresponding masks) come in different sizes. For training, however, we'll need all of them to be the same size. This can straightforwardly be accomplished by passing in `transform =` and `target_transform =` arguments. But what about data augmentation (basically always a useful measure to take)? Imagine we make use of random flipping. An input image will be flipped -- or not -- according to some probability. But if the image is flipped, the mask better had be, as well! Input and target transformations are not independent, in this case.
+
+A solution is to create a wrapper around `oxford_pet_dataset()` that lets us "hook into" the `.getitem()` method, like so:
+
+```{r}
+pet_dataset <- torch::dataset(
+  
+  inherit = oxford_pet_dataset,
+  
+  initialize = function(..., size, normalize = TRUE, augmentation = NULL) {
+    
+    self$augmentation <- augmentation
+    
+    input_transform <- function(x) {
+      x <- x %>%
+        transform_to_tensor() %>%
+        transform_resize(size) 
+      if (isTRUE(normalize)) x <- x %>%
+        transform_normalize(mean = c(0.485, 0.456, 0.406),
+                            std = c(0.229, 0.224, 0.225))
+      x
+    }
+    
+    target_transform <- function(x) {
+      x <- torch_tensor(x, dtype = torch_long())
+      x <- x[newaxis,..]
+      # interpolation = 0 makes sure we still end up with integer classes
+      x <- transform_resize(x, size, interpolation = 0)
+    }
+    
+    self$split <- split
+    
+    super$initialize(
+      ...,
+      transform = input_transform,
+      target_transform = target_transform
+    )
+    
+  },
+  .getitem = function(i) {
+    
+    item <- super$.getitem(i)
+    if (!is.null(self$augmentation)) 
+      self$augmentation(item)
+    else
+      list(x = item$x, y = item$y[1,..])
+  }
+)
+```
+
+All we have to do now is create a custom function that lets us decide on what augmentation to apply once per input-target pair, and then, manually call the respective transformation functions.
```

Comment:
Would it make sense to say "what augmentation to apply to each input-target pair"? It's simpler, but it could obscure some nuance of what's being done that I don't fully undersand.